cmd/beemo/firehose_consumer.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"strings"
+
+	comatproto "github.com/bluesky-social/indigo/api/atproto"
+	appbsky "github.com/bluesky-social/indigo/api/bsky"
+	"github.com/bluesky-social/indigo/atproto/syntax"
+	"github.com/bluesky-social/indigo/events/schedulers/autoscaling"
+	lexutil "github.com/bluesky-social/indigo/lex/util"
+
+	"github.com/bluesky-social/indigo/events"
+	"github.com/bluesky-social/indigo/repo"
+	"github.com/bluesky-social/indigo/repomgr"
+	"github.com/carlmjohnson/versioninfo"
+	"github.com/gorilla/websocket"
+)
+
+func RunFirehoseConsumer(ctx context.Context, logger *slog.Logger, relayHost string, postCallback func(context.Context, syntax.DID, syntax.RecordKey, appbsky.FeedPost) error) error {
+
+	dialer := websocket.DefaultDialer
+	u, err := url.Parse(relayHost)
+	if err != nil {
+		return fmt.Errorf("invalid relayHost URI: %w", err)
+	}
+	// always continue at the current cursor offset (don't provide cursor query param)
+	u.Path = "xrpc/com.atproto.sync.subscribeRepos"
+	logger.Info("subscribing to repo event stream", "upstream", relayHost)
+	con, _, err := dialer.Dial(u.String(), http.Header{
+		"User-Agent": []string{fmt.Sprintf("beemo/%s", versioninfo.Short())},
+	})
+	if err != nil {
+		return fmt.Errorf("subscribing to firehose failed (dialing): %w", err)
+	}
+
+	rsc := &events.RepoStreamCallbacks{
+		RepoCommit: func(evt *comatproto.SyncSubscribeRepos_Commit) error {
+			return HandleRepoCommit(ctx, logger, evt, postCallback)
+		},
+		// NOTE: could add other callbacks as needed
+	}
+
+	var scheduler events.Scheduler
```

Comment:
I would recommend using the parallelScheduler tbh, the autoscalingScheduler has been more trouble than it's worth lately.